File: Data_Structures/python/Linked_lists/Linked_lists.py
# Node class
class Node:

    # Function to initialise the node object
    def __init__(self, data):
        self.data = data
        self.next = None


# Linked List class contains a Node object
class LinkedList:

    # Function to initialize head
    def __init__(self):
        self.head = None


    def push(self, new_data):

        # 1 & 2: Allocate the Node & Put in the data
        new_node = Node(new_data)

        # 3. Make next of new Node as head
        new_node.next = self.head

        # 4. Move the head to point to new Node
        self.head = new_node


    # This function is in LinkedList class. Inserts a
    # new node after the given prev_node. This method is
    # defined inside LinkedList class shown above */
    def insertAfter(self, prev_node, new_data):

        # 1. check if the given prev_node exists
        if prev_node is None:
            print("The given previous node must inLinkedList.")
            return

        #  2. create new node & Put in the data
        new_node = Node(new_data)

        # 4. Make next of new Node as next of prev_node
        new_node.next = prev_node.next

        # 5. make next of prev_node as new_node
        prev_node.next = new_node


    # This function is defined in Linked List class
    # Appends a new node at the end.  This method is
    # defined inside LinkedList class shown above */
    def append(self, new_data):

        # 1. Create a new node
        # 2. Put in the data
        # 3. Set next as None
        new_node = Node(new_data)

        # 4. If the Linked List is empty, then make the new node as head
        if self.head is None:
            self.head = new_node
            return

        # 5. Else traverse till the last node
        last = self.head
        while (last.next):
            last = last.next

        # 6. Change the next of last node
        last.next =  new_node

    def getCount(self):
        temp = self.head
        count = 0

        while (temp):
            count += 1
            temp = temp.next
        return count

    # Given a reference to the head of a list and a key,
    # delete the first occurence of key in linked list
    def deleteNode(self, key):

        # Store head node
        temp = self.head

        # If head node itself holds the key to be deleted
        if (temp is not None):
            if (temp.data == key):
                self.head = temp.next
                temp = None
                return

        # Search for the key to be deleted, keep track of the
        # previous node as we need to change 'prev.next'
        while(temp is not None):
            if temp.data == key:
                break
            prev = temp
            temp = temp.next

        # if key was not present in linked list
        if(temp == None):
            return

        # Unlink the node from linked list
        prev.next = temp.next

        temp = None

    # Given a reference to the head of a list
    # and a position, delete the node at a given position
    def deleteNodeAtPosition(self, position):

        # If linked list is empty
        if self.head == None:
            return

        # Store head node
        temp = self.head

        # If head needs to be removed
        if position == 0:
            self.head = temp.next
            temp = None
            return

        # Find previous node of the node to be deleted
        for i in range(position -1 ):
            temp = temp.next
            if temp is None:
                break

        # If position is more than number of nodes
        if temp is None:
            return
        if temp.next is None:
            return

        # Node temp.next is the node to be deleted
        # store pointer to the next of node to be deleted
        next = temp.next.next

        # Unlink the node from linked list
        temp.next = None
        temp.next = next

    # Utility function to print the linked list
    def printList(self):
        temp = self.head
        while (temp):
            print (temp.data),
            temp = temp.next



if __name__=='__main__':

    # Start with the empty list
    llist = LinkedList()

    llist.append(6)
    llist.push(7);
    llist.push(1);
    llist.append(4)
    llist.insertAfter(llist.head.next, 8)

    print("Created linked list is:")
    llist.printList()
    print("the length of this LL is: ", llist.getCount())

    llist.deleteNode(7)
    print ("\nLinked List after Deletion of 7:")
    llist.printList()
    llist.deleteNodeAtPosition(2)
    print ("\nLinked List after Deletion of the node in position 2:")
    llist.printList()

File: Data_Structures/C++/Linked_lists/Linked_lists.cpp
#include <iostream>

using namespace std;

class Node {
    public:
        int data;
        Node *next;
};

void    printList(Node *head)
{
    while (head != NULL)
    {
        printf("%d ", head->data);
        head = head->next;
    }
}

void    push(Node **head, int data)
{
    Node *newNode = new Node();
    newNode->next = *head;
    newNode->data = data;
    *head = newNode;
}

void    insertAfter(Node *prev, int data)
{
    if (prev == NULL)
        cout << "Doesn't exist" << endl;
    Node *newNode = new Node();
    newNode->data = data;
    newNode->next = prev->next;
    prev->next = newNode;
}

void    append(Node **head, int data)
{
    Node *it = *head;
    Node *newNode = new Node();
    newNode->data = data;
    newNode->next = NULL;
    if (*head == NULL)
        *head = newNode; return ;
    while (it->next != NULL)
        it = it->next;
    it->next = newNode;
}

void    deleteNode(Node **head, int position)
{
    if (*head == NULL) return;
    Node *tmp = new Node();
    tmp = *head;
    if (position == 0)
    {
        *head = tmp->next;
        free(*head);
        return ;
    }
    for (int i = 0; tmp != NULL && i < position - 1; i++)
        tmp = tmp->next;
    if (tmp == NULL || tmp->next == NULL)
        return;
    free(tmp->next);
    Node *next = tmp->next->next;
    tmp->next = next;
}

int     size(Node *head)
{
    int i = 0;
    while (head != NULL)
    {
        head = head->next;
        i++;
    }
    return i;
}

void    swapNode(Node **head, int x, int y)
{
    if (x == y) return ;
    Node *prevX = NULL, *currentX = *head;
    while (currentX && currentX->data != x)
    {
        prevX = currentX;
        currentX = currentX->next;
    }
    Node *prevY = NULL, *currentY = *head;
    while (currentY && currentY->data != y)
    {
        prevY = currentY;
        currentY = currentY->next;
    }
    if (currentX == NULL || currentY == NULL) return ;
    // if x is not the head of linked list
    if (prevX != NULL)
        *head = currentY;
    else
        *head = currentX;

    Node *tmp = currentY->next;
    currentY->next = currentX->next;
    currentX->next = tmp;
}

int     main(void)
{
     /* Start with the empty list */
    Node* head = NULL;

    // Insert 6. So linked list becomes 6->NULL
    append(&head, 6);

    // Insert 7 at the beginning.
    // So linked list becomes 7->6->NULL
    push(&head, 7);

    // Insert 1 at the beginning.
    // So linked list becomes 1->7->6->NULL
    push(&head, 1);

    // Insert 4 at the end. So
    // linked list becomes 1->7->6->4->NULL
    append(&head, 4);

    deleteNode(&head, 4);

    // Insert 8, after 7. So linked
    // list becomes 1->7->8->6->4->NULL
    insertAfter(head->next, 8);

    cout<<"Created Linked list is: ";
    printList(head);
    cout << "size of this linked list: " << size(head);

    cout << "\n\n\n\n\n\n\n";
    reverse(head);
    printList(head);
    return (0);
}
